cfgrammar/src/lib/yacc/grammar.rs

```diff
@@ -106,17 +106,17 @@ where
     /// user defined start rule.
     pub fn new_with_storaget(yacc_kind: YaccKind, s: &str) -> YaccGrammarResult<Self> {
         let ast_validation = ast::ASTWithValidityInfo::new(yacc_kind, s);
-        Self::new_from_ast_with_validity_info(yacc_kind, ast_validation)
+        Self::new_from_ast_with_validity_info(yacc_kind, &ast_validation)
     }
 
     pub fn new_from_ast_with_validity_info(
         yacc_kind: YaccKind,
-        ast_validation: ast::ASTWithValidityInfo,
+        ast_validation: &ast::ASTWithValidityInfo,
```

Comment:
This might be a silly question, but why make this a reference given the `clone`s below?

Reply:
It is actually a key point, without making this a reference `new_from_ast_with_validity_info` takes ownership of the AST,
The `GrammarAST` however isn't moved into the `YaccGrammar`, only those portions of it which we now have to clone were moved into the `YaccGrammar`.

So, to make `GrammarAST` accessible *after* we have constructed a `YaccGrammar` from it, we need to either keep it around as a field of `YaccGrammar` or some such, alternately (as in this patch) avoid taking ownership of it herein by taking a reference to it. I *think* taking a reference to it is preferred, since that still allows you to drop it if it really is unneeded.

Reply:
> So, to make GrammarAST accessible after we have constructed a YaccGrammar from it

Is this something that e.g. the client you're writing needs to do?

Reply:
> Is this something that e.g. the client you're writing needs to do?

Yeah, but the tests added here too,
>        let grm = YaccGrammar::<u32>::new_from_ast_with_validity_info(yk, &ast_validity).unwrap();
later when we go:
>                let prod = &ast_validity.ast().prods[usize::from(pidx)];

Without taking it by reference here, The ast will have moved. Only then can we get spans from the ast.

Reply:
I wonder if rather than passing just `&` we pass a lifetime tied to `self`? So that might make it something like:

```
struct <'a> YaccGrammar {
   ast: &'a ASTWithValidityInfo
}
```

This would avoid the `clone`s (it's not so much that the `clone`s are expensive, but that it feels potentially easy to forget which `YaccGrammar` is related to which `ASTWithValidityInfo`, since it's not enforced by the static type system).

Reply:
Yeah, we could perhaps do that, but I wasn't sure if freeing the rest of the memory associated with the `GrammarAST` while keeping the `YaccGrammar` was desirable.  I could try it out though, if having these lifetimes entangled is better.

Reply:
It actually looks like entangling these lifetimes will be fairly difficult, because of the `Deserialize` trait on `YaccGrammar`.
So, you can currently serialize/deserialize a `YaccGrammar`, without a reference to an AST being involved.  I think in those cases it makes sense that we deserialize to owned values?

Reply:
Hmm. I *think* this is `Deserialize` so that `YaccGrammar` can be used by error recovery, but my memory is a bit fuzzy.

I must admit that I am slightly nervous that we're going to make the API easy to misuse -- but I don't have a quick answer for how to do any better :/

Reply:
I may very well be missing something, but I don't actually see the potential for misuse:
The biggest concern I see is that owning an AST might give the user the ability to mutate said AST.
But that has been avoided by `ASTWithValidityInfo` which only ever returns immutable references to an `&GrammarAST`.
So I don't really think this conveys anything above and beyond what can already be done besides controlling when
the AST gets dropped.

I personally, kind of think that it makes sense for the `YaccGrammar` to live beyond the length of the `GrammarAST`.
E.g. while I need to extend the lifetime of a `GrammarAST` in my program to be live until after a `StateTable` is received,
for some analysis. I don't actually need it to live beyond that, for example by the time i'm passing input strings to the grammar to be parsed I don't have any use for the `GrammarAST`.

An an alternative option (which I believe we shot down earlier) here is deriving e.g. `Clone` for `Production`, so that
you can `ast_validity.ast().prods.clone()` before `new_from_ast_with_validity_info()` takes ownership.

Reply:
> I personally, kind of think that it makes sense for the YaccGrammar to live beyond the length of the GrammarAST.

That is a fair point!